tools/src/main/scala/com/nvidia/spark/rapids/tool/profiling/AutoTuner.scala

```diff
@@ -123,212 +225,394 @@ class Config {
-            comments :+= "'spark.executor.memoryOverheadFactor' must be set " +
-              "if using 'spark.rapids.memory.pinnedPool.size'"
-          }
-        } else {
-          if (getSparkProperty(app, "spark.kubernetes.memoryOverheadFactor").isEmpty) {
-            comments :+= "'spark.kubernetes.memoryOverheadFactor' must be set " +
-              "if using 'spark.rapids.memory.pinnedPool.size'"
-          }
+  def addRecommendationForMemoryOverhead(recomValue: String): Unit = {
+    val memOverheadLookup = memoryOverheadLabel
+    appendRecommendationForMemoryMB(memOverheadLookup, recomValue)
+    getPropertyValue("spark.rapids.memory.pinnedPool.size").foreach { lookup =>
+      if (lookup != "spark.executor.memoryOverhead") {
+        if (getPropertyValue(memOverheadLookup).isEmpty) {
+          appendComment(s"'$memOverheadLookup' must be set if using " +
+            s"'spark.rapids.memory.pinnedPool.size")
         }
       }
     }
   }
 
+  def calculateRecommendations(): Unit = {
+    recommendExecutorInstances()
+    val numExecutorCores = calcNumExecutorCores
+    val execCoresExpr = () => numExecutorCores
+
+    appendRecommendation("spark.executor.cores", numExecutorCores)
+    appendRecommendation("spark.task.resource.gpu.amount",
+      calcTaskGPUAmount(execCoresExpr))
+    appendRecommendation("spark.rapids.sql.concurrentGpuTasks",
+      calcGpuConcTasks().toInt)
+    val availableMemPerExec = calcAvailableMemPerExec()
+    val availableMemPerExecExpr = () => availableMemPerExec
+    val executorHeap = calcExecutorHeap(availableMemPerExecExpr, execCoresExpr)
+    val executorHeapExpr = () => executorHeap
+    appendRecommendationForMemoryMB("spark.executor.memory", s"$executorHeap")
+    val (pinnedMemory, memoryOverhead) =
+      calcPinnedMemoryWithOverhead(executorHeapExpr, availableMemPerExecExpr)
+    appendRecommendationForMemoryMB("spark.rapids.memory.pinnedPool.size", s"$pinnedMemory")
+    addRecommendationForMemoryOverhead(s"$memoryOverhead")
+
+    recommendMaxPartitionBytes()
+    recommendShufflePartitions()
+    recommendGeneralProperties()
+  }
+
   /**
-   * Recommendation for 'spark.sql.shuffle.partitions' based on spill size.
+   * Checks whether the cluster properties are valid.
+   * If the cluster worker-info is missing entries (i.e., CPU and GPU count), it sets the entries
+   * to default values. For each default value, a comment is added to the [[comments]].
+   *
+   * @return false if the cluster properties are not loaded. e.g, all entries are set to 0.
+   *         true if the missing information were updated to default initial values.
    */
-  private def recommendShufflePartitions(recommendedConfig: Config): Unit = {
-    var shufflePartitions: Int = getSparkProperty(app, "spark.sql.shuffle.partitions")
-      .getOrElse("200").toInt
+  def processPropsAndCheck: Boolean = {
+    if (clusterProps.system.getNumCores <= 0) {
```

Comment:
we may also just want to check memory here and fallback.   Perhaps if numworkers isn't specified we default it to 1

Reply:
ideally we would add tests for all the various cases things coudl be 0